x-pack/plugins/lens/public/indexpattern_datasource/datapanel.tsx

```diff
@@ -217,88 +219,143 @@ export const InnerIndexPatternDataPanel = function InnerIndexPatternDataPanel({
     (type) => type in fieldTypeNames
   );
 
-  const displayedFields = allFields.filter((field) => {
-    if (!supportedFieldTypes.has(field.type)) {
-      return false;
-    }
+  const containsData = (field: IndexPatternField) => {
+    const fieldByName = indexBy(allFields, 'name');
+    const overallField = fieldByName[field.name];
 
-    if (
-      localState.nameFilter.length &&
-      !field.name.toLowerCase().includes(localState.nameFilter.toLowerCase())
-    ) {
-      return false;
+    return (
+      overallField && fieldExists(existingFields, currentIndexPattern.title, overallField.name)
+    );
+  };
+
+  const fieldGroups: FieldsGroup = useMemo(() => {
+    const allSupportedTypesFields = allFields.filter((field) =>
+      supportedFieldTypes.has(field.type)
+    );
+    const sorted = allSupportedTypesFields.sort(sortFields);
+    // optimization before existingFields are synced
+    if (!hasSyncedExistingFields) {
+      return {
+        ...defaultFieldGroups,
+        ...groupBy(sorted, (field) => {
+          if (field.type === 'document') {
+            return 'specialFields';
+          } else {
+            return 'emptyFields';
+          }
+        }),
+      };
     }
+    return {
+      ...defaultFieldGroups,
+      ...groupBy(sorted, (field) => {
+        if (field.type === 'document') {
+          return 'specialFields';
+        } else if (containsData(field)) {
+          return 'availableFields';
+        } else return 'emptyFields';
+      }),
+    };
+  }, [allFields, existingFields[currentIndexPattern.title], currentIndexPattern]);
 
-    if (!showEmptyFields) {
-      const indexField = currentIndexPattern && fieldByName[field.name];
-      const exists =
-        field.type === 'document' ||
-        (indexField && fieldExists(existingFields, currentIndexPattern.title, indexField.name));
-      if (localState.typeFilter.length > 0) {
-        return exists && localState.typeFilter.includes(field.type as DataType);
-      }
+  const filteredFieldGroups: FieldsGroup = useMemo(() => {
+    const filterFieldGroup = (fieldGroup: IndexPatternField[]) =>
+      fieldGroup.filter((field) => {
+        if (
+          localState.nameFilter.length &&
+          !field.name.toLowerCase().includes(localState.nameFilter.toLowerCase())
+        ) {
+          return false;
+        }
+
+        if (localState.typeFilter.length > 0) {
+          return localState.typeFilter.includes(field.type as DataType);
+        }
+        return true;
+      });
+
+    return Object.entries(fieldGroups).reduce((acc, [name, fields]) => {
+      return {
+        ...acc,
+        [name]: filterFieldGroup(fields),
+      };
+    }, defaultFieldGroups);
+  }, [fieldGroups, localState.nameFilter, localState.typeFilter]);
 
-      return exists;
+  const lazyScroll = () => {
+    if (scrollContainer) {
+      const nearBottom =
+        scrollContainer.scrollTop + scrollContainer.clientHeight >
+        scrollContainer.scrollHeight * 0.9;
```

Comment:
Scroll performance is significantly better in this update, but I would identify this function as an area of concern: it's reading DOM attributes like `scrollTop` which can be slow. It's possible that this function needs to be debounced, since scroll events on browsers like Chrome can happen within milliseconds of each other.